### PR TITLE
dev(scripts): split deployment concerns in multiple files

### PR DESCRIPTION
**Pull request by enitrat**
Splits the deployment flow in multiple files. leverage the "lazy execution"

- - -
This change is [Reviewable](https://reviewable.io/reviews/kkrt-labs/kakarot/1495)

### CODE REVIEW

File: kakarot_scripts/constants.py [hunk 1]
```diff
@@ -282,19 +282,32 @@ class ChainId(IntEnum):
...
+        "required_eth": 0.1,
+    },
+    "CreateX": {
+        "deployer": "0xeD456e05CaAb11d66C4c797dD6c1D6f9A7F352b5",
+        "signed_tx": bytes.fromhex(
+            json.loads((DATA_DIR / "signed_txs.json").read_text())["createx"]
+        ),
+        "address": "0xba5Ed099633D3B313e4D5F7bdc1305d3c28ba5Ed",
+        "required_eth": 0.3,
+    },
```
Comment (ClementWalter) [inline kakarot_scripts/constants.py:309]: since there is already a `signed_txs.json`, we can prob only do

```
PRE_EIP155_TXS = json.loads(...)
```

(possibly updating the json file)

File: kakarot_scripts/deployment/declarations.py
Comment (ClementWalter) [inline kakarot_scripts/deployment/declarations.py:1]: add the `__main__` as in the other scripts

File: kakarot_scripts/deployment/declarations.py [hunk 2]
```diff
@@ -0,0 +1,8 @@
+from kakarot_scripts.constants import DECLARED_CONTRACTS
+from kakarot_scripts.utils.starknet import declare, dump_declarations
+
+
+async def declare_contracts():
+    class_hash = {contract: await declare(contract) for contract in DECLARED_CONTRACTS}
+    dump_declarations(class_hash)
+    return class_hash
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/declarations.py:8]: is it useful somewhere?

File: kakarot_scripts/deployment/kakarot_deployment.py
Comment (ClementWalter) [inline kakarot_scripts/deployment/kakarot_deployment.py:1]: why not in the `starknet_deployment.py` and reserving the name `kakarot_deployment.py` to the EVM contracts deployed on Kakarot

File: kakarot_scripts/deployment/evm_deployments.py [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+from kakarot_scripts.utils.kakarot import get_deployments as get_evm_deployments
+from kakarot_scripts.utils.starknet import call, invoke
+
+logger = logging.getLogger(__name__)
+
+
+async def deploy_evm_contracts():
+    if not EVM_ADDRESS:
+        logger.info("ℹ️  No EVM address provided, skipping EVM deployments")
+        return
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/evm_deployments.py:16]: move into `__main__` as you did for kakaort

File: kakarot_scripts/deployment/evm_deployments.py [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+            1,
+        )
+        await invoke("kakarot", "set_coinbase", int(bridge.address, 16))
+
+    coinbase = (await call("kakarot", "get_coinbase")).coinbase
+    if coinbase == 0:
+        logger.error("❌ Coinbase is set to 0, all transaction fees will be lost")
+    else:
+        logger.info(f"✅ Coinbase set to: 0x{coinbase:040x}")
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/evm_deployments.py:46]: this should be part of a final stage called at the very end of the main deploy script where we want to make sure that the set up is correct

File: kakarot_scripts/deployment/pre_eip155_deployments.py [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            return
+
+        # lazy whitelisting of multiple contracts
+        account = await get_starknet_account()
+        register_lazy_account(account.address)
+        await whitelist_pre_eip155_contracts()
+        await execute_calls()
+        remove_lazy_account(account.address)
+
+        await deploy_pre_eip155_contracts()
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/pre_eip155_deployments.py:57]: probably not the same account deploying I don't think that it changes something to remove the lazy here

File: kakarot_scripts/deployment/evm_deployments.py (hunk 3 above)
Comment (enitrat) [reply kakarot_scripts/deployment/evm_deployments.py:27]: dont we want this to be logged also when called from the main script?

File: kakarot_scripts/deployment/kakarot_deployment.py
Comment (enitrat) [reply kakarot_scripts/deployment/kakarot_deployment.py:1]: i like to have separation for
- starknet contract deployments (miscellaneous)
- kakart deployment only
- evm contracts deployments

File: kakarot_scripts/constants.py [hunk 6]
```diff
@@ -242,6 +242,18 @@ class ChainId(IntEnum):
...
+def load_pre_eip155_transactions():
+    raw_data = json.loads((DATA_DIR / "pre_eip155_txs.json").read_text())
+
+    for tx in raw_data.values():
+        tx["signed_tx"] = bytes.fromhex(tx["signed_tx"])
+
+    return raw_data
+
+
+PRE_EIP155_TX = load_pre_eip155_transactions()
```
Comment (ClementWalter) [inline kakarot_scripts/constants.py:254]: maybe a good opportunity to introduce pydantic, which would be something like 

```
class Pre155TransactionsModel(BaseModel):
    signed_tx: bytes

    @field_validator("signed_tx", mode="before")
    def to_bytes(v):
        bytes.fromhex(...)

PRE_EIP155_TX = Pre155TransactionsModel.model_validate(json.loads((DATA_DIR / "pre_eip155_txs.json").read_text()))
```

Comment (ClementWalter) [reply kakarot_scripts/constants.py:254]: another options is to just have the `pre_eip155_txs.json` a `pre_eip155_txs.py` with directly the right types

File: kakarot_scripts/data/pre_eip155_txs.json
Comment (ClementWalter) [inline kakarot_scripts/data/pre_eip155_txs.json:1]: we can make this a python file with the right types

File: kakarot_scripts/deployment/declarations.py [hunk 7]
```diff
@@ -0,0 +1,16 @@
...
+async def declare_contracts():
+    class_hash = {contract: await declare(contract) for contract in DECLARED_CONTRACTS}
+    dump_declarations(class_hash)
+
+
+if __name__ == "__main__":
+    from uvloop import run
+
+    async def main():
+        await declare_contracts()
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/declarations.py:14]: why is this required ?

File: kakarot_scripts/deployment/declarations.py [hunk 8]
```diff
@@ -0,0 +1,16 @@
...
+from kakarot_scripts.utils.starknet import declare, dump_declarations
+
+
+async def declare_contracts():
+    class_hash = {contract: await declare(contract) for contract in DECLARED_CONTRACTS}
+    dump_declarations(class_hash)
+
+
+if __name__ == "__main__":
+    from uvloop import run
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/declarations.py:11]: for all the sub scripts, put the same final logic (with appropriate naming)

The `xxx_sync` version let us define a uv script and do `uv run xxx`

```
# %% Run
def main_sync():
    run(main())


if __name__ == "__main__":
    main_sync()
```

**Review by ClementWalter — CHANGES_REQUESTED** [review]
can you also add separator `# %%` for each section as it's more convenient to run some part on jupyter when required

File: kakarot_scripts/deployment/evm_deployments.py [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+            to_checksum_address(f'{evm_deployments["KakarotETH"]["address"]:040x}'),
+        )
+        evm_deployments["Coinbase"] = {
+            "address": int(contract.address, 16),
+            "starknet_address": contract.starknet_address,
+        }
+        await invoke("kakarot", "set_coinbase", int(contract.address, 16))
+
+    # %% Tear down
+    await execute_calls()
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/evm_deployments.py:96]: move inside the `__main__`

File: kakarot_scripts/deployment/main.py [hunk 10]
```diff
@@ -0,0 +1,72 @@
...
+logger.setLevel(logging.INFO)
+
+
+async def main():
+    # %% Declarations
+    account = await get_starknet_account()
+    register_lazy_account(account.address)
+    logger.info(f"ℹ️  Using account 0x{account.address:064x} as deployer")
+    balance_before = await get_balance(account.address)
+    await declare_contracts()
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/main.py:34]:
```suggestion

    # %% Account initialization
    account = await get_starknet_account()
    register_lazy_account(account.address)
    logger.info(f"ℹ️  Using account 0x{account.address:064x} as deployer")
    balance_before = await get_balance(account.address)
    
    # %% Declarations    
    await declare_contracts()
```

File: kakarot_scripts/deployment/main.py [hunk 11]
```diff
@@ -0,0 +1,72 @@
...
+    await deploy_evm_contracts()
+    await execute_calls()
+
+    # Needs whitelist tx to be executed first
+    await deploy_pre_eip155_contracts()
+    await execute_calls()
+    remove_lazy_account(account.address)
+
+    # Must be sequential
+    await deploy_messaging_contracts()
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/main.py:52]:
```suggestion

    # Must be sequential
    remove_lazy_account(account.address)    
    await deploy_messaging_contracts()
```

File: kakarot_scripts/deployment/messaging_deployments.py [hunk 12]
```diff
@@ -0,0 +1,107 @@
...
+    from uvloop import run
+
+    async def main():
+        try:
+            await RPC_CLIENT.get_class_hash_at(get_starknet_deployments()["kakarot"])
+        except Exception:
+            logger.error("❌ Kakarot is not deployed, exiting...")
+            return
+        await deploy_l1_contracts()
+        await deploy_messaging_contracts()
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/messaging_deployments.py:105]: why not splitting this in two functions, naming could be

```suggestion
        await deploy_l1_messaging_contracts()
        await deploy_l2_messaging_contracts()
```

File: kakarot_scripts/deployment/pre_eip155_deployments.py (hunk 5 above)
Comment (ClementWalter) [inline kakarot_scripts/deployment/pre_eip155_deployments.py:57]: define in the main module and not inside the if, add the sync version

File: kakarot_scripts/utils/kakarot.py [hunk 13]
```diff
@@ -747,29 +745,97 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+
+#     response = await send_pre_eip155_transaction(
+#         deployer_evm_address, deployer_starknet_address, signed_tx, max_fee
+#     )
+
+#     receipt, response, success, gas_used = response
+#     deployed_address = response[1]
+#     logger.info(f"✅ {name} Deployed at: 0x{deployed_address:040x}")
+#     deployed_starknet_address = await get_starknet_address(deployed_address)
+#     return {"address": deployed_address, "starknet_address": deployed_starknet_address}
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:773]: remove or uncomment?

File: kakarot_scripts/utils/kakarot.py [hunk 14]
```diff
@@ -747,29 +745,97 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+
+
+async def whitelist_pre_eip155_tx(name: str) -> bool:
+    tx_instance = PRE_EIP155_TX[name]
+    deployer_evm_address = tx_instance["deployer"]
+    tx_instance["address"]
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:779]:
```suggestion
```

File: kakarot_scripts/utils/kakarot.py [hunk 15]
```diff
@@ -747,29 +745,97 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+    signed_tx = tx_instance["signed_tx"]
+    amount = tx_instance["required_eth"]
+
+    # check the nonce of the deployer for an early return if it's not 0.
+    # Either the nonce is 0, or the account should not be deployed already.
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:784]:
```suggestion
    # Either the nonce is 0, or the account is already deployed.
```

File: kakarot_scripts/utils/kakarot.py [hunk 16]
```diff
@@ -747,29 +745,97 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
     if nonce != 0:
         logger.info(
             f"ℹ️  Nonce for {deployer_evm_address} is not 0 ({nonce}), skipping transaction"
         )
         return
 
-    response = await send_pre_eip155_transaction(
-        deployer_evm_address, deployer_starknet_address, signed_tx, max_fee
+    # Deploy and fund deployer for the authorization callback
+    await deploy_and_fund_evm_address(deployer_evm_address, amount)
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:810]: i don't think that deploying the deployer belongs to the whitelist process, move this to `send_pre_eip155_transaction`

File: kakarot_scripts/utils/kakarot.py [hunk 17]
```diff
@@ -747,29 +745,97 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+#     )
+
+#     receipt, response, success, gas_used = response
+#     deployed_address = response[1]
+#     logger.info(f"✅ {name} Deployed at: 0x{deployed_address:040x}")
+#     deployed_starknet_address = await get_starknet_address(deployed_address)
+#     return {"address": deployed_address, "starknet_address": deployed_starknet_address}
+
+
+async def whitelist_pre_eip155_tx(name: str) -> bool:
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:776]: so the whitelist function should
- analyze the tx to get the corresponding nonce
- call kakarot to get the deployer nonce
- skip if tx nonce < kakarot nonce
- invoke authorized otherwise

no eoa deployment here

File: kakarot_scripts/utils/kakarot.py [hunk 18]
```diff
@@ -747,29 +745,97 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+            )
+        ).tx_count
+    except Exception as e:
+        if (
+            f"Requested contract address {deployer_starknet_address} is not deployed"
+            in str(e.data)
+        ):
+            nonce = 0
+        else:
+            raise e
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:802]: this should be better put in a kakarot utils

```
def eth_get_transaction_count(evm_address):
    # ...
```

File: kakarot_scripts/utils/kakarot.py [hunk 19]
```diff
@@ -747,29 +745,97 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+def get_unsigned_encoded_tx_data(signed_tx: bytes):
+    rlp_decoded = rlp.decode(signed_tx)
+    unsigned_tx_data = rlp_decoded[:-3]
+    unsigned_encoded_tx = rlp.encode(unsigned_tx_data)
+    return unsigned_encoded_tx
+
+
+def get_signature(signed_tx: bytes) -> Tuple[int, int, int]:
+    rlp_decoded = rlp.decode(signed_tx)
+    return rlp_decoded[-3:]
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:841]: this transaction decoding functions are specific to pre 155, I'd keep them in the `send_pre_eip155_transaction`

we can also do a significant refactoring of all the eth related modules of the project to use the exec specs but it's another topic

File: kakarot_scripts/utils/kakarot.py (hunk 16 above)
Comment (enitrat) [reply kakarot_scripts/utils/kakarot.py:810]: I need to deploy the deployer to call `set_authorized_pre_eip155_tx` on it.

File: kakarot_scripts/deployment/evm_deployments.py [hunk 20]
```diff
@@ -0,0 +1,121 @@
+import logging
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/evm_deployments.py:1]:
```suggestion
# %% Imports
import logging
```

File: kakarot_scripts/deployment/evm_deployments.py [hunk 21]
```diff
@@ -0,0 +1,121 @@
...
+from kakarot_scripts.utils.kakarot import dump_deployments as dump_evm_deployments
+from kakarot_scripts.utils.kakarot import get_deployments as get_evm_deployments
+from kakarot_scripts.utils.starknet import call, execute_calls
+from kakarot_scripts.utils.starknet import get_deployments as get_starknet_deployments
+from kakarot_scripts.utils.starknet import invoke
+
+logger = logging.getLogger(__name__)
+
+
+async def deploy_evm_contracts():
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/evm_deployments.py:25]:
```suggestion
# %% Main
async def deploy_evm_contracts():
    # %% Tokens
```

Comment (ClementWalter) [reply kakarot_scripts/deployment/evm_deployments.py:25]: sandwiching the. `def` let you run the cells easily

File: kakarot_scripts/deployment/evm_deployments.py [hunk 22]
```diff
@@ -0,0 +1,121 @@
...
+# %% Run
+async def main():
+    try:
+        await RPC_CLIENT.get_class_hash_at(get_starknet_deployments()["kakarot"])
+    except Exception:
+        logger.error("❌ Kakarot is not deployed, exiting...")
+        return
+
+    if not EVM_ADDRESS:
+        logger.info("ℹ️  No EVM address provided, skipping EVM deployments")
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/evm_deployments.py:109]:
```suggestion
        logger.warn("⚠️  No EVM address provided, skipping EVM deployments")
```

File: kakarot_scripts/deployment/kakarot_deployment.py
Comment (ClementWalter) [inline kakarot_scripts/deployment/kakarot_deployment.py:1]: add block separators

File: kakarot_scripts/deployment/kakarot_deployment.py [hunk 23]
```diff
@@ -0,0 +1,91 @@
...
+    get_starknet_account,
+    invoke,
+    register_lazy_account,
+    remove_lazy_account,
+)
+
+logger = logging.getLogger(__name__)
+
+
+async def deploy_or_upgrade_kakarot(account):
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/kakarot_deployment.py:28]:
```suggestion
async def deploy_or_upgrade_kakarot(owner):
```

File: kakarot_scripts/deployment/kakarot_deployment.py [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+                "kakarot",
+                "set_cairo1_helpers_class_hash",
+                class_hash["Cairo1Helpers"],
+            )
+        else:
+            logger.info("✅ Kakarot already up to date.")
+    else:
+        starknet_deployments["kakarot"] = await deploy_starknet(
+            "kakarot",
+            account.address,  # owner
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/kakarot_deployment.py:55]:
```suggestion
            owner.address,  # owner
```

File: kakarot_scripts/deployment/pre_eip155_deployments.py [hunk 25]
```diff
@@ -0,0 +1,65 @@
...
+    try:
+        await RPC_CLIENT.get_class_hash_at(get_deployments()["kakarot"])
+    except Exception:
+        logger.error("❌ Kakarot is not deployed, exiting...")
+        return
+
+    account = await get_starknet_account()
+    register_lazy_account(account.address)
+    await whitelist_pre_eip155_contracts()
+    await execute_calls()
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/pre_eip155_deployments.py:53]: batch all the authorizations in one multicall

File: kakarot_scripts/deployment/starknet_deployments.py [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+def main_sync():
+    run(main())
+
+
+async def main():
+    try:
+        await RPC_CLIENT.get_class_by_hash(get_declarations()["kakarot"])
+    except Exception:
+        logger.error("❌ Classes were not declared, exiting...")
+        return
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/starknet_deployments.py:73]:
```suggestion
```

File: kakarot_scripts/utils/kakarot.py [hunk 27]
```diff
@@ -517,11 +527,23 @@ async def eth_get_code(address: Union[int, str]):
...
-        await _call_starknet("account_contract", "get_nonce", address=starknet_address)
-    ).nonce
+async def eth_get_transaction_count(evm_address, starknet_address=None):
+    kakarot_contract = _get_starknet_contract("kakarot")
+    try:
+        nonce = (
+            await kakarot_contract.functions["eth_get_transaction_count"].call(
+                int(evm_address, 16)
+            )
+        ).tx_count
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:537]:
```suggestion
        nonce = (
            await call("kakarot", "eth_get_transaction_count", int(evm_address, 16))
        ).tx_count
```

File: kakarot_scripts/utils/kakarot.py [hunk 28]
```diff
@@ -517,11 +527,23 @@ async def eth_get_code(address: Union[int, str]):
     )
 
 
-async def eth_get_transaction_count(address: Union[int, str]):
-    starknet_address = await get_starknet_address(address)
-    return (
-        await _call_starknet("account_contract", "get_nonce", address=starknet_address)
-    ).nonce
+async def eth_get_transaction_count(evm_address, starknet_address=None):
+    kakarot_contract = _get_starknet_contract("kakarot")
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:531]:
```suggestion
```

File: kakarot_scripts/utils/kakarot.py [hunk 29]
```diff
@@ -517,11 +527,23 @@ async def eth_get_code(address: Union[int, str]):
     )
 
 
-async def eth_get_transaction_count(address: Union[int, str]):
-    starknet_address = await get_starknet_address(address)
-    return (
-        await _call_starknet("account_contract", "get_nonce", address=starknet_address)
-    ).nonce
+async def eth_get_transaction_count(evm_address, starknet_address=None):
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:530]:
```suggestion
async def eth_get_transaction_count(evm_address):
```

File: kakarot_scripts/utils/kakarot.py [hunk 30]
```diff
@@ -517,11 +527,23 @@ async def eth_get_code(address: Union[int, str]):
...
+    kakarot_contract = _get_starknet_contract("kakarot")
+    try:
+        nonce = (
+            await kakarot_contract.functions["eth_get_transaction_count"].call(
+                int(evm_address, 16)
+            )
+        ).tx_count
+    except Exception as e:
+        if f"Requested contract address {starknet_address} is not deployed" in str(
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:539]:
```suggestion
        starknet_address = await call("kakarot", "get_starknet_address", int(evm_address, 16))
        if f"Requested contract address {starknet_address} is not deployed" in str(
```

File: kakarot_scripts/utils/kakarot.py [hunk 31]
```diff
@@ -676,6 +698,7 @@ async def deploy_and_fund_evm_address(evm_address: str, amount: float):
     """
     Deploy an EOA linked to the given EVM address and fund it with amount ETH.
     """
+    logger.info(f"ℹ️  Funding EVM address {evm_address}")
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:701]: `eth_balance_of` is not funding addresses

File: kakarot_scripts/utils/kakarot.py [hunk 32]
```diff
@@ -747,29 +770,27 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+    tx_instance["signed_tx"]
+    amount = tx_instance["required_eth"]
+
+    # check the nonce of the deployer for an early return if it's not 0.
+    # Either the nonce is 0, or the account is already deployed.
+    deployer_starknet_address = (
+        f"0x{await get_starknet_address(deployer_evm_address):064x}"
+    )
+    nonce = await eth_get_transaction_count(
+        deployer_evm_address, deployer_starknet_address
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:785]:
```suggestion
        deployer_evm_address
```

File: kakarot_scripts/utils/kakarot.py [hunk 33]
```diff
@@ -747,29 +770,27 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+    tx_instance = PRE_EIP155_TX[name]
+    deployer_evm_address = tx_instance["deployer"]
+    tx_instance["signed_tx"]
+    amount = tx_instance["required_eth"]
+
+    # check the nonce of the deployer for an early return if it's not 0.
+    # Either the nonce is 0, or the account is already deployed.
+    deployer_starknet_address = (
+        f"0x{await get_starknet_address(deployer_evm_address):064x}"
+    )
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:783]:
```suggestion
```

File: kakarot_scripts/utils/kakarot.py [hunk 34]
```diff
@@ -747,29 +770,27 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
+    nonce = await eth_get_transaction_count(
+        deployer_evm_address, deployer_starknet_address
     )
-    nonce = (
-        await _call_starknet(
-            "account_contract", "get_nonce", address=deployer_starknet_address
-        )
-    ).nonce
     if nonce != 0:
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:787]: compare to tx nonce and not 0

File: kakarot_scripts/utils/kakarot.py [hunk 35]
```diff
@@ -747,29 +770,27 @@ async def store_bytecode(bytecode: Union[str, bytes], **kwargs):
...
-    )
+    # Deploy and fund deployer to enable the authorization callback when calling set_authorized_pre_eip155_tx
+    await deploy_and_fund_evm_address(deployer_evm_address, amount)
 
-    receipt, response, success, gas_used = response
-    deployed_address = response[1]
-    logger.info(f"✅ {name} Deployed at: 0x{deployed_address:040x}")
-    deployed_starknet_address = await get_starknet_address(deployed_address)
-    return {"address": deployed_address, "starknet_address": deployed_starknet_address}
+    tx_instance["should_deploy"] = True
```
Comment (ClementWalter) [inline kakarot_scripts/utils/kakarot.py:796]:
```suggestion
    tx_instance["deployed"] = True
```

File: kakarot_scripts/deployment/declarations.py [hunk 36]
```diff
@@ -0,0 +1,24 @@
+# %% Imports
+from uvloop import run
+
+from kakarot_scripts.constants import DECLARED_CONTRACTS
+from kakarot_scripts.utils.starknet import declare, dump_declarations
+
+# %%
+
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/declarations.py:9]:
```suggestion
```

File: kakarot_scripts/deployment/declarations.py [hunk 37]
```diff
@@ -0,0 +1,24 @@
...
+    dump_declarations(class_hash)
+
+
+# %% Run
+def main_sync():
+    run(declare_contracts())
+
+
+# %%
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/declarations.py:22]:
```suggestion
```

File: kakarot_scripts/deployment/kakarot_deployment.py [hunk 38]
```diff
@@ -0,0 +1,96 @@
...
+    invoke,
+    register_lazy_account,
+    remove_lazy_account,
+)
+
+logger = logging.getLogger(__name__)
+
+# %%
+
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/kakarot_deployment.py:31]:
```suggestion
```

File: kakarot_scripts/deployment/kakarot_deployment.py [hunk 39]
```diff
@@ -0,0 +1,96 @@
+# %% Imports
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/kakarot_deployment.py:2]:
```suggestion
```

File: kakarot_scripts/deployment/kakarot_deployment.py [hunk 40]
```diff
@@ -0,0 +1,96 @@
...
+    remove_lazy_account,
+)
+
+logger = logging.getLogger(__name__)
+
+# %%
+
+
+async def deploy_or_upgrade_kakarot(owner):
+    # %% Deploy
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/kakarot_deployment.py:33]:
```suggestion
    # %% Get data
```

File: kakarot_scripts/deployment/main.py [hunk 41]
```diff
@@ -0,0 +1,83 @@
+import logging
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/main.py:1]:
```suggestion
# %% Imports
import logging
```

File: kakarot_scripts/deployment/main.py [hunk 42]
```diff
@@ -0,0 +1,83 @@
...
+    get_starknet_account,
+    register_lazy_account,
+    remove_lazy_account,
+)
+
+logging.basicConfig()
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/main.py:31]:
```suggestion
# %%
```

File: kakarot_scripts/deployment/messaging_deployments.py [hunk 43]
```diff
@@ -0,0 +1,116 @@
...
+)
+from kakarot_scripts.utils.starknet import call
+from kakarot_scripts.utils.starknet import get_deployments as get_starknet_deployments
+from kakarot_scripts.utils.starknet import invoke
+
+logger = logging.getLogger(__name__)
+
+# %%
+
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/messaging_deployments.py:26]:
```suggestion
```

File: kakarot_scripts/deployment/pre_eip155_deployments.py [hunk 44]
```diff
@@ -0,0 +1,80 @@
...
+# %%
+
+
+async def deploy_pre_eip155_senders():
+    # %% Deployers
+    for contract_name in PRE_EIP155_TX.keys():
+        await deploy_pre_eip155_sender(contract_name)
+    # %%
+
+
```
Comment (ClementWalter) [inline kakarot_scripts/deployment/pre_eip155_deployments.py:33]:
```suggestion


# %%
```

**Review by ClementWalter — APPROVED** [review]
